oce does not build on ppc64le aarch

Currently oce does not pass tests on new POWER architecture with the following error:

Errors while running CTest
48/48 Test #28: BRepMeshTestSuite.testMeshTorus .................   Passed    2.61 sec

98% tests passed, 1 tests failed out of 48

Total Test time (real) =   3.15 sec

The following tests FAILED:
	 24 - BRepOffsetAPITestSuite.testEvolvedShape (Failed)
make[1]: *** [test] Error 8
dh_auto_test: make -j8 test ARGS+=-j8 returned exit code 2
make: *** [build-arch] Error 2
Makefile:140: recipe for target 'test' failed
make[1]: Leaving directory '/«PKGBUILDDIR»/obj-powerpc64le-linux-gnu'
debian/rules:14: recipe for target 'build-arch' failed

We would like to have it ported and full building in ppc64el architecture

I've been playing with this for the last few days, and I'm having trouble understanding what the problem is exactly.

Following the instructions in [INSTALL.Unix](https://github.com/tpaviot/oce/blob/master/INSTALL.Unix) I'm able to run the following commands without any error:
- Creating a build dir, and running **cmake ..** inside it
- Running **make**
- Running **make install**

When it comes down to testing however, running **make test** gives me the following results:

```
hernan@debianle:~/desarrollo/oce/build$ make test
Running tests...
Test project /home/desarrollo/oce/build
      Start  1: BRepOffsetAPITestSuite.testEvolvedShape
 1/48 Test  #1: BRepOffsetAPITestSuite.testEvolvedShape .........***Failed    0.03 sec
      Start  2: BRepOffsetAPITestSuite.testOffsetBox
 2/48 Test  #2: BRepOffsetAPITestSuite.testOffsetBox ............   Passed    0.04 sec
      Start  3: IGESImportTestSuite.testImportIGES_1
 3/48 Test  #3: IGESImportTestSuite.testImportIGES_1 ............   Passed    0.36 sec
      Start  4: gpTestSuite.testgp_PntConstruct
 4/48 Test  #4: gpTestSuite.testgp_PntConstruct .................   Passed    0.01 sec
      Start  5: gpTestSuite.testgp_PntDistance
 5/48 Test  #5: gpTestSuite.testgp_PntDistance ..................   Passed    0.01 sec
      Start  6: gpTestSuite.testBuildThousandsPoints
 6/48 Test  #6: gpTestSuite.testBuildThousandsPoints ............   Passed    0.01 sec
      Start  7: gpTestSuite.testGCC42gp_Trsf
 7/48 Test  #7: gpTestSuite.testGCC42gp_Trsf ....................   Passed    0.01 sec
      Start  8: OCAFExportTestSuite.testExportAscii
 8/48 Test  #8: OCAFExportTestSuite.testExportAscii .............   Passed    0.04 sec
      Start  9: OCAFExportTestSuite.testExportNonAscii
 9/48 Test  #9: OCAFExportTestSuite.testExportNonAscii ..........   Passed    0.03 sec
      Start 10: OCAFExportTestSuite.testOverflow
10/48 Test #10: OCAFExportTestSuite.testOverflow ................   Passed    0.03 sec
      Start 11: BRepMeshTestSuite.testMeshBox
11/48 Test #11: BRepMeshTestSuite.testMeshBox ...................   Passed    0.02 sec
      Start 12: BRepMeshTestSuite.testMeshSphere
12/48 Test #12: BRepMeshTestSuite.testMeshSphere ................   Passed    0.26 sec
      Start 13: BRepMeshTestSuite.testMeshTorus
13/48 Test #13: BRepMeshTestSuite.testMeshTorus .................   Passed    3.45 sec
      Start 14: BRepImportTestSuite.testImportBRep1
14/48 Test #14: BRepImportTestSuite.testImportBRep1 .............   Passed    0.33 sec
      Start 15: BRepImportTestSuite.testImportBRep2
15/48 Test #15: BRepImportTestSuite.testImportBRep2 .............   Passed    0.03 sec
      Start 16: BRepImportTestSuite.testImportBRep3
16/48 Test #16: BRepImportTestSuite.testImportBRep3 .............   Passed    0.03 sec
      Start 17: BRepImportTestSuite.testImportBRep4
17/48 Test #17: BRepImportTestSuite.testImportBRep4 .............   Passed    0.05 sec
      Start 18: TCollectionTestSuite.testSequenceOfPnt
18/48 Test #18: TCollectionTestSuite.testSequenceOfPnt ..........   Passed    0.13 sec
      Start 19: Standard_StdAllocatorTestSuite.testTraits
19/48 Test #19: Standard_StdAllocatorTestSuite.testTraits .......   Passed    0.01 sec
      Start 20: Standard_StdAllocatorTestSuite.testContainers
20/48 Test #20: Standard_StdAllocatorTestSuite.testContainers ...   Passed    0.01 sec
      Start 21: BRepExtremaTestSuite.testDistSphereSphere
21/48 Test #21: BRepExtremaTestSuite.testDistSphereSphere .......   Passed    0.01 sec
      Start 22: BRepExtremaTestSuite.testDistSphereVertex
22/48 Test #22: BRepExtremaTestSuite.testDistSphereVertex .......   Passed    0.01 sec
      Start 23: BRepExtremaTestSuite.testDistVertexEdge
23/48 Test #23: BRepExtremaTestSuite.testDistVertexEdge .........   Passed    0.01 sec
      Start 24: BRepExtremaTestSuite.testDistEdgeVertex
24/48 Test #24: BRepExtremaTestSuite.testDistEdgeVertex .........   Passed    0.01 sec
      Start 25: BRepExtremaTestSuite.testDistVertexFace
25/48 Test #25: BRepExtremaTestSuite.testDistVertexFace .........   Passed    0.01 sec
      Start 26: BRepExtremaTestSuite.testDistFaceVertex
26/48 Test #26: BRepExtremaTestSuite.testDistFaceVertex .........   Passed    0.01 sec
      Start 27: BRepGPropTestSuite.testComputeBoxVolume
27/48 Test #27: BRepGPropTestSuite.testComputeBoxVolume .........   Passed    0.01 sec
      Start 28: BRepGPropTestSuite.testComputeBoxSurface
28/48 Test #28: BRepGPropTestSuite.testComputeBoxSurface ........   Passed    0.01 sec
      Start 29: BRepGPropTestSuite.testComputeSphereVolume
29/48 Test #29: BRepGPropTestSuite.testComputeSphereVolume ......   Passed    0.01 sec
      Start 30: STEPImportTestSuite.testImportAP203_1
30/48 Test #30: STEPImportTestSuite.testImportAP203_1 ...........   Passed    0.23 sec
      Start 31: STEPImportTestSuite.testImportAP203_2
31/48 Test #31: STEPImportTestSuite.testImportAP203_2 ...........   Passed    0.34 sec
      Start 32: STEPImportTestSuite.testImportAP214_1
32/48 Test #32: STEPImportTestSuite.testImportAP214_1 ...........   Passed    0.44 sec
      Start 33: STEPImportTestSuite.testImportAP214_2
33/48 Test #33: STEPImportTestSuite.testImportAP214_2 ...........   Passed    0.17 sec
      Start 34: STEPImportTestSuite.testImportAP214_3
34/48 Test #34: STEPImportTestSuite.testImportAP214_3 ...........   Passed    0.10 sec
      Start 35: SortToolsTestSuite.testQuickSortOfInteger
35/48 Test #35: SortToolsTestSuite.testQuickSortOfInteger .......   Passed    0.00 sec
      Start 36: BRepPrimAPITestSuite.testCreateBox
36/48 Test #36: BRepPrimAPITestSuite.testCreateBox ..............   Passed    0.01 sec
      Start 37: BRepPrimAPITestSuite.testCreateFlatBox
37/48 Test #37: BRepPrimAPITestSuite.testCreateFlatBox ..........   Passed    0.01 sec
      Start 38: BRepPrimAPITestSuite.testCreateSmallestBox
38/48 Test #38: BRepPrimAPITestSuite.testCreateSmallestBox ......   Passed    0.01 sec
      Start 39: UnitsAPITestSuite.testCurrentUnits
39/48 Test #39: UnitsAPITestSuite.testCurrentUnits ..............   Passed    0.01 sec
      Start 40: UnitsAPITestSuite.testMDTVCurrentUnits
40/48 Test #40: UnitsAPITestSuite.testMDTVCurrentUnits ..........   Passed    0.05 sec
      Start 41: BRepAlgoAPITestSuite.testCutBox
41/48 Test #41: BRepAlgoAPITestSuite.testCutBox .................   Passed    0.04 sec
      Start 42: BRepAlgoAPITestSuite.testCutCylSphere
42/48 Test #42: BRepAlgoAPITestSuite.testCutCylSphere ...........   Passed    0.03 sec
      Start 43: TDataXtdTestSuite.testPattern
43/48 Test #43: TDataXtdTestSuite.testPattern ...................   Passed    0.03 sec
      Start 44: TestSuite.testNullPointer
44/48 Test #44: TestSuite.testNullPointer .......................   Passed    0.00 sec
      Start 45: TestSuite.testFloatEq
45/48 Test #45: TestSuite.testFloatEq ...........................   Passed    0.00 sec
      Start 46: TestSuite.testFloatNeq
46/48 Test #46: TestSuite.testFloatNeq ..........................   Passed    0.00 sec
      Start 47: TestSuite.testBoolean
47/48 Test #47: TestSuite.testBoolean ...........................   Passed    0.00 sec
      Start 48: TestSuite.testIntegerLighter
48/48 Test #48: TestSuite.testIntegerLighter ....................   Passed    0.00 sec

98% tests passed, 1 tests failed out of 48

Total Test time (real) =   6.46 sec

The following tests FAILED:
          1 - BRepOffsetAPITestSuite.testEvolvedShape (Failed)
Errors while running CTest
Makefile:137: recipe for target 'test' failed
make: *** [test] Error 8
```

Note that while the failing test is the same, the error message and number is different from the one @leitao posted.

I'm able to _build and install_ oce on my setup (with ppc64le arch), but this test keeps failing. Is it right to assume that this issue will be closed when the test is fixed, then?

Yes, I'm consistenly seeing the following error:
```
The following tests FAILED:
          1 - BRepOffsetAPITestSuite.testEvolvedShape (Failed)
Errors while running CTest
Makefile:137: recipe for target 'test' failed
make: *** [test] Error 8
```

No randomized order. I've already performed the build proces a couple of times (with and without the flag -DOCE_TESTING:BOOL=ON) and the test order is always the same.

I doubt it fails because a previous test leaving the environment unstable. At least in my case, BRepOffsetAPITestSuite.testEvolvedShape is always the first test that runs (notice that there's another test inside the same suit, and it passes).

The most I've been able to do is track down [the place where it blows up to this line](https://github.com/tpaviot/oce/blob/master/test/BRepOffsetAPI_test/BRepOffsetAPI_test.cpp#L19).
I know that because I've been modifying, rebuilding and running the tests several times, commenting out every line and replacing
```
ASSERT_TRUE(S.IsDone());
```
with
```
ASSERT_EQ (1,1);
```
on each run.

If the test is run with this code, then it passes:
```
TEST(BRepOffsetAPITestSuite, testEvolvedShape)
{
    BRepBuilderAPI_MakePolygon P;
    P.Add(gp_Pnt(0.,0.,0.));
    P.Add(gp_Pnt(200.,0.,0.));
    P.Add(gp_Pnt(200.,200.,0.));
    P.Add(gp_Pnt(0.,200.,0.));
    P.Add(gp_Pnt(0.,0.,0.));
    BRepBuilderAPI_MakePolygon wprof(gp_Pnt(0.,0.,0.),gp_Pnt(-60.,-60.,-200.));
    //BRepOffsetAPI_MakeEvolved S(P.Wire(),wprof.Wire(),GeomAbs_Arc,Standard_True,Standard_False,Standard_True,0.0001);
    //S.Build();
    //ASSERT_TRUE(S.IsDone());
    ASSERT_EQ(1,1);
}
```
(notice all the commented lines)

And if you change that code just a little, like this:
```
TEST(BRepOffsetAPITestSuite, testEvolvedShape)
{
    BRepBuilderAPI_MakePolygon P;
    P.Add(gp_Pnt(0.,0.,0.));
    P.Add(gp_Pnt(200.,0.,0.));
    P.Add(gp_Pnt(200.,200.,0.));
    P.Add(gp_Pnt(0.,200.,0.));
    P.Add(gp_Pnt(0.,0.,0.));
    BRepBuilderAPI_MakePolygon wprof(gp_Pnt(0.,0.,0.),gp_Pnt(-60.,-60.,-200.));
    BRepOffsetAPI_MakeEvolved S(P.Wire(),wprof.Wire(),GeomAbs_Arc,Standard_True,Standard_False,Standard_True,0.0001);
    //S.Build();
    //ASSERT_TRUE(S.IsDone());
    ASSERT_EQ(1,1);
}
```
and run only this test with the following command
```
env GTEST_CATCH_EXCEPTIONS=0 ctest -R testEvolvedShape -V
```

then you get
```
hernan@debianle:~/desarrollo/oce/build$ env GTEST_CATCH_EXCEPTIONS=0 ctest -R testEvolvedShape -V
UpdateCTestConfiguration  from :/home/hernan/desarrollo/oce/build/DartConfiguration.tcl
Parse Config file:/home/hernan/desarrollo/oce/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/hernan/desarrollo/oce/build/DartConfiguration.tcl
Parse Config file:/home/hernan/desarrollo/oce/build/DartConfiguration.tcl
Test project /home/hernan/desarrollo/oce/build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: BRepOffsetAPITestSuite.testEvolvedShape

1: Test command: /home/hernan/desarrollo/oce/build/test/BRepOffsetAPI_test/BRepOffsetAPI_test "--gtest_filter=BRepOffsetAPITestSuite.testEvolvedShape"
1: Test timeout computed to be: 1500
1: Note: Google Test filter = BRepOffsetAPITestSuite.testEvolvedShape
1: [==========] Running 1 test from 1 test case.
1: [----------] Global test environment set-up.
1: [----------] 1 test from BRepOffsetAPITestSuite
1: [ RUN      ] BRepOffsetAPITestSuite.testEvolvedShape
1: terminate called after throwing an instance of 'Standard_NotImplemented'
1/1 Test #1: BRepOffsetAPITestSuite.testEvolvedShape ...***Exception: Other  0.03 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.04 sec

The following tests FAILED:
          1 - BRepOffsetAPITestSuite.testEvolvedShape (OTHER_FAULT)
Errors while running CTest

```
So appearently, calling 
```
BRepOffsetAPI_MakeEvolved S(P.Wire(),wprof.Wire(),GeomAbs_Arc,Standard_True,Standard_False,Standard_True,0.0001);
``` 
throws an Standard_NotImplemented exception.

I still couldn't figure out why, but I believe that's what causing the test to fail.

--- a/src/Adaptor3d/Adaptor3d_SurfaceOfRevolution.cxx
+++ b/src/Adaptor3d/Adaptor3d_SurfaceOfRevolution.cxx
@@ -547,7 +547,7 @@ GeomAbs_SurfaceType Adaptor3d_SurfaceOfRevolution::GetType() const
   //
   switch ( myBasisCurve->GetType()) {
   case GeomAbs_Line:    {
-    const gp_Ax1& Axe = (myBasisCurve->Line()).Position();
+    const gp_Ax1 Axe = (myBasisCurve->Line()).Position();
     
     if (myAxis.IsParallel(Axe, TolAng)) {
       bRet=GeomAbs_Cylinder;

